torch/testing/_internal/common_device_type.py

```diff
@@ -205,46 +212,83 @@ def _get_precision_override(self, test, dtype):
     # Creates device-specific tests.
     @classmethod
     def instantiate_test(cls, name, test):
-        test_name = name + "_" + cls.device_type
 
-        dtypes = cls._get_dtypes(test)
-        if dtypes is None:  # Test has no dtype variants
-            assert not hasattr(cls, test_name), "Redefinition of test {0}".format(test_name)
+        def instantiate_test_helper(cls, name, *, test, dtype, op):
 
-            @wraps(test)
-            def instantiated_test(self, test=test):
-                device_arg = cls.get_primary_device() if not hasattr(test, 'num_required_devices') else cls.get_all_devices()
-                return test(self, device_arg)
-
-            setattr(cls, test_name, instantiated_test)
-        else:  # Test has dtype variants
-            for dtype in dtypes:
-                # Constructs dtype suffix
+            # Constructs the test's name
+            test_name = name
+            if op is not None:
+                test_name += "_" + op.name
+            test_name += "_" + cls.device_type
+            if dtype is not None:
                 if isinstance(dtype, (list, tuple)):
-                    dtype_str = ""
                     for d in dtype:
-                        dtype_str += "_" + str(d).split('.')[1]
+                        test_name += "_" + str(d).split('.')[1]
```

Comment:
this is not the fault of this PR, but I find the fact that you have to do this to extract out `int8` (e.g.) really annoying

Reply:
If it's any consolation I think the PR is a net improvement on the quality of this code.